### PR TITLE
CSHARP-5416: GetType not equals comparison translates incorrectly to MQL.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstSimplifier.cs [hunk 1]
```diff
@@ -364,6 +364,16 @@ static AstExpression UltimateGetFieldInput(AstGetFieldExpression getField)
             }
         }
 
+        public override AstNode VisitNotFilterOperation(AstNotFilterOperation node)
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstSimplifier.cs:367]: Simplifies `{ f : { $not : { $exists : false } } }` to `{ f : { $exists : true } }`

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs [hunk 2]
```diff
@@ -30,28 +29,49 @@ namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToFilter
...
-        public static bool CanTranslate(Expression leftExpression, Expression rightExpression)
+        public static bool CanTranslate(
+            Expression leftExpression,
+            AstComparisonFilterOperator comparisonOperator,
+            Expression rightExpression)
         {
             return
                 leftExpression is MethodCallExpression methodCallExpression &&
                 methodCallExpression.Method.Is(ObjectMethod.GetType) &&
+                (comparisonOperator == AstComparisonFilterOperator.Eq || comparisonOperator == AstComparisonFilterOperator.Ne) &&
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs:40]: Before we were assuming that operator was always `==` but that was wrong.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs [hunk 3]
```diff
@@ -30,28 +29,49 @@ namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToFilter
...
                 rightExpression is ConstantExpression;
         }
 
-        public static AstFilter Translate(TranslationContext context, BinaryExpression expression, MethodCallExpression getTypeExpression, Expression typeConstantExpression)
+        public static AstFilter Translate(
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs:44]: Lots of these diffs are just whitespace differences due to indentation changes... :(

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs [hunk 4]
```diff
@@ -30,28 +29,49 @@ namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToFilter
...
+                var filter = discriminatorConvention switch
+                {
+                    IHierarchicalDiscriminatorConvention hierarchicalDiscriminatorConvention => DiscriminatorAstFilter.TypeEquals(discriminatorField, hierarchicalDiscriminatorConvention, nominalType, actualType),
+                    _ => DiscriminatorAstFilter.TypeEquals(discriminatorField, discriminatorConvention, nominalType, actualType),
+                };
+
+                return comparisonOperator switch
+                {
+                    AstComparisonFilterOperator.Eq => filter,
+                    AstComparisonFilterOperator.Ne => AstFilter.Not(filter),
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs:69]: This is the main change to support `!=`

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs [hunk 5]
```diff
@@ -30,28 +29,49 @@ namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToFilter
...
+                    _ => throw new ExpressionNotSupportedException(expression, because: $"comparison operator {comparisonOperator} is not supported")
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslator.cs:70]: We should never reach this line but switch requires all cases to be covered.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Driver.Linq;
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Linq.Linq3Implementation.Translators.ExpressionToFilterTranslators.ExpressionTranslators
+{
+    public class GetTypeComparisonExpressionToFilterTranslatorTests : Linq3IntegrationTest
+    {
+        [Fact]
+        public void Hierarchical_documents_should_be_serialized_as_expected()
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/GetTypeComparisonExpressionToFilterTranslatorTests.cs:28]: We need to test against both hierarchical and scalar discriminators.

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request merged**